### PR TITLE
Support for local kind cluster

### LINKED ISSUE #152: Local kind cluster installation
I would like to install the project in a kind cluster for local testing.

### PR DESCRIPTION
**Pull request by nstogner**
Add local kind support.

[See Diagram](https://raw.githubusercontent.com/substratusai/substratus/5c813780ccc3d139dd12586869dd85e3e9898fab/docs/diagrams/arch-kind.excalidraw.png)

Fixes #152

### CODE REVIEW

File: internal/sci/kind/server.go
Comment (brandonjbjelland) [inline internal/sci/kind/server.go:56]: Reaching a bit here but is it better to `os.Stat(dst)` the destination before copying so as to not overwrite?

Comment (nstogner) [reply internal/sci/kind/server.go:56]: I think overwriting the stored tarball is what we do in the real buckets today

Comment (brandonjbjelland) [reply internal/sci/kind/server.go:80]: > overwriting the stored tarball is what we do in the real buckets today

Huh - I bet we have a not-so-subtle bug on GCP where one user could easily overwrite another's before the builder can start a build against it.

Comment (nstogner) [reply internal/sci/kind/server.go:80]: Are you thinking two users could both be running `kubectl notebook -d .` on the same Notebook at the same time?

Comment (nstogner) [reply internal/sci/kind/server.go:80]: If so, we are not currently accounting for that. Notebooks are intended to be used by 1 user at a time at the moment.

File: internal/sci/kind/server.go
Comment (samos123) [inline internal/sci/kind/server.go:18]: what is the reason for doing this? Trying to understand why this is needed

Comment (nstogner) [reply internal/sci/kind/server.go:18]: Not needed, just creates a compile-time error if the `Server{}` does not satisfy the intended `sci.ControllerServer` interface.

File: config/install-kind/namespace.yaml
Comment (samos123) [inline config/install-kind/namespace.yaml:1]: I think this should be included by default in all installation not just kind?

File: config/sci-kind/deployment.yaml
Comment (samos123) [inline config/sci-kind/deployment.yaml:1]: rename to `deployment_patch.yaml` to follow same pattern as with `manager_patch.yaml`

File: config/sci-kind/service.yaml
Comment (samos123) [inline config/sci-kind/service.yaml:1]: rename to `service_patch.yaml`

**Review by samos123 — CHANGES_REQUESTED** [review]

File: cmd/sci-kind/main.go
Comment (nstogner) [reply cmd/sci-kind/main.go:27]: Agreed, I have already forgotten myself. Will document better

File: Makefile
Comment (nstogner) [reply Makefile:159]: Thought was: we might want a `dev-run-{cloud}` and `dev-skaffold-{cloud}` for each cloud. `dev-run-*` currently runs locally... There are probably some better names... Let me know your thoughts.

File: Makefile
Comment (nstogner) [reply Makefile:164]: Good callout

File: Makefile
Comment (nstogner) [reply Makefile:286]: Lets merge your PR 1st

File: internal/cloud/kind.go
Comment (samos123) [inline internal/cloud/kind.go:28]: Add auto configure for ArtifactURL instead of setting it in configmap

File: config/install-kind/namespace.yaml
Comment (samos123) [reply config/install-kind/namespace.yaml:1]: I have done this in my PR/branch for now as well: https://github.com/substratusai/substratus/pull/185/commits/01d2ec10fd41dbf126348534888411eb0f43c1b6

Approach I took is add namespace.yaml under config/manager. I thought it wasn't worth creating a separate directory.

File: Makefile
Comment (samos123) [reply Makefile:159]: in that case I like your naming, I like the convention of `dev-skaffold-{CLOUD}`. I guess the issue for kind is there is no way to do `dev-run-kind`? Maybe `dev-run-kind` just points to `dev-skaffold-kind`.

File: skaffold.kind.yaml
Comment (samos123) [inline skaffold.kind.yaml:16]: how does this work, does kind automatically use the images built on the host? This shouldn't be working unless you do this: https://kind.sigs.k8s.io/docs/user/quick-start/#loading-an-image-into-your-cluster

Comment (samos123) [reply skaffold.kind.yaml:16]: found out why it works: https://skaffold.dev/docs/environment/local-cluster/

File: Makefile
Comment (samos123) [reply Makefile:286]: I am now in favor of merging my PR first. I tested it very thoroughly and feel it's ready. Let's discuss more offline.

Comment (brandonjbjelland) [reply Makefile:286]: Yeah it's become necessary at this point. I don't think we'll end up in this state in the future (developing multiple SCIs at once) but I think we can avoid this if we worked in smaller chunks as PRs (SCI, SCI implementation, deployment bits, and lastly the controller integration - in that order)

File: config/install-kind/manager_patch.yaml
Comment (brandonjbjelland) [reply config/install-kind/manager_patch.yaml:7]: Yeah holding off until your PR is merged.

**Review by samos123 — APPROVED** [review]

**Pull request merged**
